torch: Make __version__ better with comparisons

seemethere opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61556 torch: Make __version__ better with comparisons**

Prior to 1.10.0 `torch.__version__` was stored as a str and so many did
comparisons against `torch.__version__` as if it were a str. In order to not
break them we have TorchVersion which masquerades as a str while also
having the ability to compare against both packaging.version.Version as
well as tuples of values, eg. (1, 2, 1)

Examples:
  Comparing a TorchVersion object to a Version object
```
TorchVersion('1.10.0a') > Version('1.10.0a')
```
  Comparing a TorchVersion object to a Tuple object
```
TorchVersion('1.10.0a') > (1, 2)    # 1.2
TorchVersion('1.10.0a') > (1, 2, 1) # 1.2.1
```

  Comparing a TorchVersion object against a string
```
TorchVersion('1.10.0a') > '1.2'
TorchVersion('1.10.0a') > '1.2.1'
```

Fixes https://github.com/pytorch/pytorch/issues/61540

Signed-off-by: Eli Uriegas <eliuriegas@fb.com>

Differential Revision: [D29671234](https://our.internmc.facebook.com/intern/diff/D29671234)

seemethere committed:
torch: Make __version__ better with comparisons

Prior to 1.10.0 torch.__version__ was stored as a str and so many did
comparisons against torch.__version__ as if it were a str. In order to not
break them we have TorchVersion which masquerades as a str while also
having the ability to compare against both packaging.version.Version as
well as tuples of values, eg. (1, 2, 1)

Examples:
  Comparing a TorchVersion object to a Version object
  >>> TorchVersion('1.10.0a') > Version('1.10.0a')

  Comparing a TorchVersion object to a Tuple object
  >>> TorchVersion('1.10.0a') > (1, 2)    # 1.2
  >>> TorchVersion('1.10.0a') > (1, 2, 1) # 1.2.1

  Comparing a TorchVersion object against a string
  >>> TorchVersion('1.10.0a') > '1.2'
  >>> TorchVersion('1.10.0a') > '1.2.1'

Signed-off-by: Eli Uriegas <eliuriegas@fb.com>

[ghstack-poisoned]

seemethere committed:
Update on "torch: Make __version__ better with comparisons"


Prior to 1.10.0 `torch.__version__` was stored as a str and so many did
comparisons against `torch.__version__` as if it were a str. In order to not
break them we have TorchVersion which masquerades as a str while also
having the ability to compare against both packaging.version.Version as
well as tuples of values, eg. (1, 2, 1)

Examples:
  Comparing a TorchVersion object to a Version object
```
TorchVersion('1.10.0a') > Version('1.10.0a')
```
  Comparing a TorchVersion object to a Tuple object
```
TorchVersion('1.10.0a') > (1, 2)    # 1.2
TorchVersion('1.10.0a') > (1, 2, 1) # 1.2.1
```

  Comparing a TorchVersion object against a string
```
TorchVersion('1.10.0a') > '1.2'
TorchVersion('1.10.0a') > '1.2.1'
```

Resolves https://github.com/pytorch/pytorch/issues/61540

Signed-off-by: Eli Uriegas <eliuriegas@fb.com>

[ghstack-poisoned]

seemethere committed:
Update on "torch: Make __version__ better with comparisons"


Prior to 1.10.0 `torch.__version__` was stored as a str and so many did
comparisons against `torch.__version__` as if it were a str. In order to not
break them we have TorchVersion which masquerades as a str while also
having the ability to compare against both packaging.version.Version as
well as tuples of values, eg. (1, 2, 1)

Examples:
  Comparing a TorchVersion object to a Version object
```
TorchVersion('1.10.0a') > Version('1.10.0a')
```
  Comparing a TorchVersion object to a Tuple object
```
TorchVersion('1.10.0a') > (1, 2)    # 1.2
TorchVersion('1.10.0a') > (1, 2, 1) # 1.2.1
```

  Comparing a TorchVersion object against a string
```
TorchVersion('1.10.0a') > '1.2'
TorchVersion('1.10.0a') > '1.2.1'
```

Resolves https://github.com/pytorch/pytorch/issues/61540

Signed-off-by: Eli Uriegas <eliuriegas@fb.com>

[ghstack-poisoned]

seemethere committed:
Update on "torch: Make __version__ better with comparisons"


Prior to 1.10.0 `torch.__version__` was stored as a str and so many did
comparisons against `torch.__version__` as if it were a str. In order to not
break them we have TorchVersion which masquerades as a str while also
having the ability to compare against both packaging.version.Version as
well as tuples of values, eg. (1, 2, 1)

Examples:
  Comparing a TorchVersion object to a Version object
```
TorchVersion('1.10.0a') > Version('1.10.0a')
```
  Comparing a TorchVersion object to a Tuple object
```
TorchVersion('1.10.0a') > (1, 2)    # 1.2
TorchVersion('1.10.0a') > (1, 2, 1) # 1.2.1
```

  Comparing a TorchVersion object against a string
```
TorchVersion('1.10.0a') > '1.2'
TorchVersion('1.10.0a') > '1.2.1'
```

Resolves https://github.com/pytorch/pytorch/issues/61540

Signed-off-by: Eli Uriegas <eliuriegas@fb.com>

[ghstack-poisoned]

seemethere committed:
Update on "torch: Make __version__ better with comparisons"


Prior to 1.10.0 `torch.__version__` was stored as a str and so many did
comparisons against `torch.__version__` as if it were a str. In order to not
break them we have TorchVersion which masquerades as a str while also
having the ability to compare against both packaging.version.Version as
well as tuples of values, eg. (1, 2, 1)

Examples:
  Comparing a TorchVersion object to a Version object
```
TorchVersion('1.10.0a') > Version('1.10.0a')
```
  Comparing a TorchVersion object to a Tuple object
```
TorchVersion('1.10.0a') > (1, 2)    # 1.2
TorchVersion('1.10.0a') > (1, 2, 1) # 1.2.1
```

  Comparing a TorchVersion object against a string
```
TorchVersion('1.10.0a') > '1.2'
TorchVersion('1.10.0a') > '1.2.1'
```

Resolves https://github.com/pytorch/pytorch/issues/61540

Signed-off-by: Eli Uriegas <eliuriegas@fb.com>

[ghstack-poisoned]

seemethere commented on .jenkins/pytorch/win-test-helpers/build_pytorch.bat:
```diff
@@ -22,7 +22,7 @@ call %INSTALLER_DIR%\install_miniconda3.bat
 
 :: Install ninja and other deps
-if "%REBUILD%"=="" ( pip install -q "ninja==1.10.0.post1" dataclasses typing_extensions "expecttest==0.1.3" )
+if "%REBUILD%"=="" ( pip install -q "ninja==1.10.0.post1" dataclasses typing_extensions "expecttest==0.1.3" packaging )
```
Needed for https://github.com/pytorch/pytorch/blob/ded4ca5f152f237d3b91df8d0412508dda79fe4d/.jenkins/pytorch/win-test-helpers/build_pytorch.bat#L129-L130

malfet replied:
This is no longer needed, is it?

seemethere commented:
@seemethere has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29671234).

seemethere committed:
Update on "torch: Make __version__ better with comparisons"


Prior to 1.10.0 `torch.__version__` was stored as a str and so many did
comparisons against `torch.__version__` as if it were a str. In order to not
break them we have TorchVersion which masquerades as a str while also
having the ability to compare against both packaging.version.Version as
well as tuples of values, eg. (1, 2, 1)

Examples:
  Comparing a TorchVersion object to a Version object
```
TorchVersion('1.10.0a') > Version('1.10.0a')
```
  Comparing a TorchVersion object to a Tuple object
```
TorchVersion('1.10.0a') > (1, 2)    # 1.2
TorchVersion('1.10.0a') > (1, 2, 1) # 1.2.1
```

  Comparing a TorchVersion object against a string
```
TorchVersion('1.10.0a') > '1.2'
TorchVersion('1.10.0a') > '1.2.1'
```

Resolves https://github.com/pytorch/pytorch/issues/61540

Signed-off-by: Eli Uriegas <eliuriegas@fb.com>

Differential Revision: [D29671234](https://our.internmc.facebook.com/intern/diff/D29671234)

[ghstack-poisoned]

seemethere commented on docs/source/conf.py:
```diff
@@ -147,14 +147,15 @@
 project = 'PyTorch'
 copyright = '2019, Torch Contributors'
 author = 'Torch Contributors'
+torch_version = str(torch.__version__)
```
To fix CI failures related to:

```
make: *** [html-stable] Error 2
++ code=2
++ '[' 2 -ne 0 ']'
++ set +x
=========================
WARNING: The config value `version' has type `TorchVersion', defaults to `str'.
WARNING: The config value `release' has type `TorchVersion', defaults to `str'.
=========================
Docs build failed. If the failure is not clear, scan back in the log
for any WARNINGS or for the line build finished with problems
(tried to echo the WARNINGS above the ==== line)
=========================
++ return 2
```

See: https://github.com/pytorch/pytorch/runs/3052078756?check_suite_focus=true

malfet commented on torch/torch_version.py:
```diff
@@ -0,0 +1,64 @@
+            #   * (1, 20, 1)  -> Version("1.20.1")
+            return Version('.'.join((str(item) for item in inp)))
+        else:
+            raise InvalidVersion(inp)
```
```suggestion
            raise packaging.version.InvalidVersion(inp)
```

malfet commented on torch/torch_version.py:
```diff
@@ -0,0 +1,64 @@
+    def __gt__(self, cmp):
+        try:
+            return Version(self).__gt__(self._convert_to_version(cmp))
+        except InvalidVersion:
```
```suggestion
        except packaging.version.InvalidVersion:
```

malfet reviewed (changes requested):
Please use copy of version package that is shipped with every Python installation rather that introduce new dependency. See https://setuptools.readthedocs.io/en/latest/pkg_resources.html

seemethere committed:
Update on "torch: Make __version__ better with comparisons"


Prior to 1.10.0 `torch.__version__` was stored as a str and so many did
comparisons against `torch.__version__` as if it were a str. In order to not
break them we have TorchVersion which masquerades as a str while also
having the ability to compare against both packaging.version.Version as
well as tuples of values, eg. (1, 2, 1)

Examples:
  Comparing a TorchVersion object to a Version object
```
TorchVersion('1.10.0a') > Version('1.10.0a')
```
  Comparing a TorchVersion object to a Tuple object
```
TorchVersion('1.10.0a') > (1, 2)    # 1.2
TorchVersion('1.10.0a') > (1, 2, 1) # 1.2.1
```

  Comparing a TorchVersion object against a string
```
TorchVersion('1.10.0a') > '1.2'
TorchVersion('1.10.0a') > '1.2.1'
```

Fixes https://github.com/pytorch/pytorch/issues/61540

Signed-off-by: Eli Uriegas <eliuriegas@fb.com>

Differential Revision: [D29671234](https://our.internmc.facebook.com/intern/diff/D29671234)

[ghstack-poisoned]

seemethere committed:
Update on "torch: Make __version__ better with comparisons"


Prior to 1.10.0 `torch.__version__` was stored as a str and so many did
comparisons against `torch.__version__` as if it were a str. In order to not
break them we have TorchVersion which masquerades as a str while also
having the ability to compare against both packaging.version.Version as
well as tuples of values, eg. (1, 2, 1)

Examples:
  Comparing a TorchVersion object to a Version object
```
TorchVersion('1.10.0a') > Version('1.10.0a')
```
  Comparing a TorchVersion object to a Tuple object
```
TorchVersion('1.10.0a') > (1, 2)    # 1.2
TorchVersion('1.10.0a') > (1, 2, 1) # 1.2.1
```

  Comparing a TorchVersion object against a string
```
TorchVersion('1.10.0a') > '1.2'
TorchVersion('1.10.0a') > '1.2.1'
```

Fixes https://github.com/pytorch/pytorch/issues/61540

Signed-off-by: Eli Uriegas <eliuriegas@fb.com>

Differential Revision: [D29671234](https://our.internmc.facebook.com/intern/diff/D29671234)

[ghstack-poisoned]

seemethere committed:
Update on "torch: Make __version__ better with comparisons"


Prior to 1.10.0 `torch.__version__` was stored as a str and so many did
comparisons against `torch.__version__` as if it were a str. In order to not
break them we have TorchVersion which masquerades as a str while also
having the ability to compare against both packaging.version.Version as
well as tuples of values, eg. (1, 2, 1)

Examples:
  Comparing a TorchVersion object to a Version object
```
TorchVersion('1.10.0a') > Version('1.10.0a')
```
  Comparing a TorchVersion object to a Tuple object
```
TorchVersion('1.10.0a') > (1, 2)    # 1.2
TorchVersion('1.10.0a') > (1, 2, 1) # 1.2.1
```

  Comparing a TorchVersion object against a string
```
TorchVersion('1.10.0a') > '1.2'
TorchVersion('1.10.0a') > '1.2.1'
```

Fixes https://github.com/pytorch/pytorch/issues/61540

Signed-off-by: Eli Uriegas <eliuriegas@fb.com>

Differential Revision: [D29671234](https://our.internmc.facebook.com/intern/diff/D29671234)

[ghstack-poisoned]

seemethere commented on torch/torch_version.py:
```diff
@@ -0,0 +1,64 @@
+from functools import total_ordering
+from typing import Iterable, Union
+
+from packaging.version import InvalidVersion, Version  # type: ignore[import]
```
Done

z-a-f commented:
How are the cuda versions handled? Such as "1.9.0+cu102"

z-a-f commented on torch/torch_version.py:
```diff
@@ -0,0 +1,68 @@
+    def _convert_to_version(self, inp: Union[packaging.version.Version, str, Iterable]) -> packaging.version.Version:
+        if isinstance(inp, Version):
+            return inp
+        elif isinstance(inp, str):
+            return Version(inp)
```
I feel like this could potentially mask some useful information (s.a. "1.9.0+cu102"). Also, I think the packaging.Version returns `LegacyVersion` for cases like "torch-deploy-1.8"

seemethere replied:
Actually for cases like `torch-deploy-1.8` it'd raise an `InvalidVersion` exception thus causing `TorchVersion` to fall back into regular string comparison.

```

In [3]: Version('torch-deploy-1.8')
---------------------------------------------------------------------------
InvalidVersion                            Traceback (most recent call last)
<ipython-input-3-c4aa93e1d2e3> in <module>
----> 1 Version('torch-deploy-1.8')

~/work/pytorch-stack/.venv/lib/python3.8/site-packages/pkg_resources/_vendor/packaging/version.py in __init__(self, version)
    275         match = self._regex.search(version)
    276         if not match:
--> 277             raise InvalidVersion("Invalid version: '{0}'".format(version))
    278 
    279         # Store the parsed out pieces of the version

InvalidVersion: Invalid version: 'torch-deploy-1.8'
```

For cases like `1.9.0+cu102` the information about the `+local` version will still remain in the `.local` function.

```
In [6]: v = Version('1.9.0+cu102')

In [7]: v.major
Out[7]: 1

In [8]: v.minor
Out[8]: 9

In [9]: v.micro
Out[9]: 0

In [10]: v.post

In [11]: v.local
Out[11]: 'cu102'
```

'+local' versioning is respected even when comparing against a string:
```
In [14]: from torch.torch_version import TorchVersion

In [15]: TorchVersion('1.9.0+cu102') == '1.9.0+cu102'
Out[15]: True
```

z-a-f replied:
Got it. There is another little snag -- do we want to compare the local versions? In my opinion, if the user has "1.9.0+cu101", it shouldn't matter if it is cuda or not, right? If there is a check such as `TorchVersion("1.9.0+cu102") == "1.9.0"`, I think it should return True. wdyt?

seemethere commented:
@seemethere has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29671234).

malfet commented on torch/torch_version.py:
```diff
@@ -0,0 +1,68 @@
+            TorchVersion('1.10.0a') > '1.2.1'
+    """
+    # fully qualified type names here to appease mypy
+    def _convert_to_version(self, inp: Union[packaging.version.Version, str, Iterable]) -> packaging.version.Version:
```
As you are creating shortcut on line 7, referencing version by full name is no longer neeeded, is it?
```suggestion
    def _convert_to_version(self, inp: Union[Version, str, Iterable]) -> Version:
```

seemethere replied:
Unfortunately I had to use the fully qualified name here since `mypy` didn't like it when I used the short hand version

seemethere replied:
Without this mypy actually fails see:
```
.venv ❯ mypy --config=mypy.ini torch/torch_version.py
torch/torch_version.py:35:46: error: Variable "torch.torch_version.Version" is not valid as a type  [valid-type]
torch/torch_version.py:35:46: note: See https://mypy.readthedocs.io/en/latest/common_issues.html#variables-vs-type-aliases
Found 1 error in 1 file (checked 1 source file)
```

<details>

<summary> Diff </summary>

```diff
diff --git a/torch/torch_version.py b/torch/torch_version.py
index bb01294aeb..e987452711 100644
--- a/torch/torch_version.py
+++ b/torch/torch_version.py
@@ -32,7 +32,7 @@ class TorchVersion(str):
             TorchVersion('1.10.0a') > '1.2.1'
     """
     # fully qualified type names here to appease mypy
-    def _convert_to_version(self, inp: Union[packaging.version.Version, str, Iterable]) -> packaging.version.Version:
+    def _convert_to_version(self, inp: Union[Version, str, Iterable]) -> Version:
         if isinstance(inp, Version):
             return inp
         elif isinstance(inp, str):
```

</details>

Context: https://mypy.readthedocs.io/en/latest/common_issues.html#variables-vs-type-aliases

seemethere committed:
Update on "torch: Make __version__ better with comparisons"


Prior to 1.10.0 `torch.__version__` was stored as a str and so many did
comparisons against `torch.__version__` as if it were a str. In order to not
break them we have TorchVersion which masquerades as a str while also
having the ability to compare against both packaging.version.Version as
well as tuples of values, eg. (1, 2, 1)

Examples:
  Comparing a TorchVersion object to a Version object
```
TorchVersion('1.10.0a') > Version('1.10.0a')
```
  Comparing a TorchVersion object to a Tuple object
```
TorchVersion('1.10.0a') > (1, 2)    # 1.2
TorchVersion('1.10.0a') > (1, 2, 1) # 1.2.1
```

  Comparing a TorchVersion object against a string
```
TorchVersion('1.10.0a') > '1.2'
TorchVersion('1.10.0a') > '1.2.1'
```

Fixes https://github.com/pytorch/pytorch/issues/61540

Signed-off-by: Eli Uriegas <eliuriegas@fb.com>

Differential Revision: [D29671234](https://our.internmc.facebook.com/intern/diff/D29671234)

[ghstack-poisoned]

seemethere commented:
@seemethere has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29671234).
